### PR TITLE
fix(kds): send NACK only when resource is invalid and do not retry

### PR DESCRIPTION
**Pull request by lukidzi**
### Checklist prior to review

Based on the investigation we noticed that our force retrier might cause some issues when connected with the sleep. We retried once the resource was NACKed, but it's incorrect xDS behavior. If the client returns NACK it means that the resource is incorrect and shouldn't be synced until is fixed. Now instead of retrying NACK we decided to distinguish validation errors from storage errors. Once there is a storage error we restart a whole component and on the initial stream sync all the resources, but for the validation error, we wait until the resource is fixed, and once it's changed we try to apply it again.

* Do not set owner reference for synced resource has been already done https://github.com/kumahq/kuma/blob/master/pkg/plugins/runtime/k8s/webhooks/owner_reference_mutator.go#L88 and should be backported to 2.6 and 2.7(probably)
* remove force retrier and add schema validation
* added test that validate that once resource is fixed, it's synced to the client 
* restart component on the db error, NACK on validation

- [ ] [Link to relevant issue][1] as well as docs and UI issues -- fix: https://github.com/kumahq/kuma/issues/10463
- [ ] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [ ] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [ ] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [ ] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

**Comment by lukidzi** [comment]
Should be backported to 2.6 and 2.7

File: pkg/kds/zone/components_test.go [hunk 1]
```diff
@@ -303,6 +303,48 @@ var _ = Describe("Zone Sync", func() {
...
+			err := globalStore.Create(context.Background(), &mesh.MeshResource{Spec: invalidMesh}, store.CreateByKey("mesh-1", model.NoMesh))
+			Expect(err).ToNot(HaveOccurred())
+
+			// should not be synchronized
+			Consistently(func(g Gomega) {
+				actual := mesh.MeshResourceList{}
+				err := zoneStore.List(context.Background(), &actual)
+				g.Expect(err).ToNot(HaveOccurred())
+				g.Expect(actual.Items).To(BeEmpty())
+			}, "5s", "100ms").Should(Succeed())
```
[the remark this answers is not in the document]
Comment (lukidzi) [reply pkg/kds/zone/components_test.go]: Currently, we are unable to, we would need to rework the unit test setup to add stats callback.

File: pkg/kds/v2/client/kds_client.go [hunk 2]
```diff
@@ -75,6 +87,16 @@ func (s *kdsSyncClient) Receive() error {
 			}
 			return errors.Wrap(err, "failed to receive a discovery response")
 		}
+
+		if err := received.Validate(); err != nil {
+			if err := s.kdsStream.NACK(received.Type, err); err != nil {
```
Comment (lobkovilya) [inline pkg/kds/v2/client/kds_client.go:92]: please restore the log `s.log.Info("error during callback received, sending NACK", "err", err)`, it's useful for debugging

Comment (lobkovilya) [reply pkg/kds/v2/client/kds_client.go:92]: ah but the text should be different

File: pkg/kds/zone/components_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/kds/zone/components_test.go]: I see, given the fact we reconcile the snapshot every 100ms in unit tests, probably 1s will be enough

File: pkg/kds/global/components_test.go [hunk 3]
```diff
@@ -70,6 +71,38 @@ var _ = Describe("Global Sync", func() {
...
+							mesh_proto.ServiceTag:  "web",
+							mesh_proto.ProtocolTag: "http",
+						},
+					},
+				},
+			},
+		}
+		err := zoneStores[0].Create(context.Background(), &mesh.DataplaneResource{Spec: dp}, store.CreateByKey("dp-1", "mesh-1"))
+		Expect(err).ToNot(HaveOccurred())
+		Eventually(func() int {
```
Comment (jakubdyszkiewicz) [inline pkg/kds/global/components_test.go:97]: Does Eventually make sense in this context? Shouldn't this be Consistentnly?

File: pkg/kds/v2/client/kds_client.go [hunk 4]
```diff
@@ -75,6 +87,17 @@ func (s *kdsSyncClient) Receive() error {
...
+			s.log.Info("received resource is invalid, sending NACK", "err", err)
+			if err := s.kdsStream.NACK(received.Type, err); err != nil {
+				if err == io.EOF {
+					return nil
+				}
+				return errors.Wrap(err, "failed to NACK a discovery response")
+			}
+			continue
+		}
 		s.log.V(1).Info("DeltaDiscoveryResponse received", "response", received)
```
Comment (jakubdyszkiewicz) [inline pkg/kds/v2/client/kds_client.go:101]: I think this log should go above the change with validation that you added

File: pkg/kds/global/components_test.go [hunk 5]
```diff
@@ -70,6 +71,38 @@ var _ = Describe("Global Sync", func() {
...
+			actual := mesh.DataplaneResourceList{}
+			err := globalStore.List(context.Background(), &actual)
+			Expect(err).ToNot(HaveOccurred())
```
Comment (jakubdyszkiewicz) [inline pkg/kds/global/components_test.go:100]: nit: I'd inject Gomega to eventually if you use Expect. Just like you did for components_test

**Review by jakubdyszkiewicz — APPROVED** [review]
a couple of nits, but looks good for me 👍

**Pull request merged**